### PR TITLE
Harmonize Treatment of Host Header in Pedant

### PR DESCRIPTION
**Pull request by lbakerchef**
Harmonize treatment of Host header in Pedant.

Chef::ServerAPI always sets the Host header to HOSTNAME:PORT.
We do the same here for consistency in order to avoid sigv4
signing issues.

verify:
https://buildkite.com/chef/chef-chef-server-master-verify/builds/2323#95b54d19-1e87-45a2-9092-3e69ceb99fb7

adhoc:
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/1605#5a0aca9e-5a4e-49df-a761-cf18f0a27042

### CODE REVIEW

File: oc-chef-pedant/spec/api/cookbooks/read_spec.rb
Comment (stevendanna) [inline oc-chef-pedant/spec/api/cookbooks/read_spec.rb:331]: I know it is silly, but maybe we can pull this out to a utility function so that we can just have this comment in one place.  Also, the comment could do more to explain what the signing issue is.  For instance:

`Chef::ServerAPI always sets the Host header to HOSTNAME:PORT. We do the same thing here since the Version 4 AWS Signature scheme uses the Host header as part of the canonically signed content.`

Also, am I correct that this isn't strictly necessary since we are also doing the Host-header fallback handling in bookshelf for compatibility?

Comment (lbakerchef) [reply oc-chef-pedant/spec/api/cookbooks/read_spec.rb:331]: It is strictly necessary in the case of using S3 vs bookshelf.

Comment (lbakerchef) [reply oc-chef-pedant/spec/api/cookbooks/read_spec.rb:331]: Will investigate pulling this out to a utility function.

File: oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb
Comment (markan) [inline oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb:335]: How far back does that behavior go? I think it's been in there a while, but it would be worth noting that if we being stricter than we used to. Pedant has been used as a guide for third party API libraries, so they may have incorporated the behavior. This is *not* intended to block this PR, but more of a thing we will have to note when the V4 signing work merges, and probably call out in the release notes, as it might break things (even if they are unsupported, its always nice to note when the contract changes).

**Review by markan — APPROVED** [review]

File: oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb
Comment (stevendanna) [reply oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb:335]: Yup, the V4 signing is definitely going to break some clients unfortunately.  When I looked at the issue last, I don't see a way around it for S3-backed chef-server.  For bookshelf-backed chef-servers we have some flexibility because we can do the auth check using both headers and accept either.

Comment (stevendanna) [reply oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb:335]: Note that I _think_ all supported versions of Chef client should be OK but we should double check that, understand how far back users can be while still working, and document the issue for 3rd party clients.

Comment (stevendanna) [reply oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb:335]: Note, when I say "break some clients" I mean clients broadly.  I think all supported versions of Chef will be OK, but we'll definitely want to document it for 3rd party clients and confirm which versions of Chef-Client work with v4 signing.

Comment (lbakerchef) [reply oc-chef-pedant/lib/pedant/rspec/cookbook_util.rb:335]: Just to provide this FYI, I did 'hand testing' of the last 3-4 Chef clients a while back (manual knife cookbook uploads and downloads and what not).  Those all passed for the testing I did.

File: oc-chef-pedant/spec/api/cookbooks/read_spec.rb
Comment (lbakerchef) [reply oc-chef-pedant/spec/api/cookbooks/read_spec.rb]: Factored-out things into a utility function.  Let me know if it's what you had in mind.

**Review by stevendanna — APPROVED** [review]

**Comment by stevendanna** [comment]
Please clean up the commits before merging, but this looks good to me!

**Comment by lbakerchef** [comment]
Thanks!

**Review by PrajaktaPurohit — APPROVED** [review]

**Pull request merged**
